### PR TITLE
Helpers for querying DB

### PR DESCRIPTION
**Pull request by am-kantox**
First of all, let me thank you for `ex_cldr`, `ex_money`, and for the hugely valuable contribution to _Elixir_ ecosystem in general.

This PR has a low value but still might be helpful for newcomers to better express their intent. `fragment/1` is fragile in many ways, so helpers like these might simplify the queries _and_ make them more readable.

```elixir
Organization
|> where([o], money_is(o.payroll, Money.new!(100, :USD)))
|> select([o], o.payroll)
|> Repo.all()
#⇒ [Money.new(:USD, "100"), Money.new(:USD, "100")]

Organization
|> where([o], amount_ge(o.payroll, 90))
|> select([o], o.payroll)
|> Repo.all()
#⇒ [Money.new(:AUD, "90"), Money.new(:USD, "100")]

Organization
|> where([o], o.name == ^"Lemon Inc.")
|> total_by([o], o.payroll, :USD)
|> Repo.one()
#⇒ [Money.new(:USD, "210")]
```

The architecture allows easy implementation of `MySQL` adapter (via [`JSON_EXTRACT`](https://dev.mysql.com/doc/refman/8.0/en/json-search-functions.html#function_json-extract),) please let me know if you are interested in adding all that to the library.

Thanks again, and have a nice day.

### CODE REVIEW

**Comment by kipcole9** [comment]
@am-kantox, thank you, that's much appreciated and a very good idea. I would prefer to have symmetry between the two types if at all possible (composite and map). Possible?  And I think `money_is` would be more appropriately named `money_eq` to be more consistent with Elixirs `:eq`, `:lt`, `:gt` symbols.

**Comment by kipcole9** [comment]
And thank you for the kind words, its appreciated and it helps keep my motivation going!

**Comment by am-kantox** [comment]
> I would prefer to have symmetry between the two types if at all possible (composite and map). Possible?

Yes, and no. As I said, it’s possible to implement the adapter for `MySQL` and I am all in into implementing it. As it is now, it _still_ allows you to plug in the home-baked adapter with `use Money.Ecto.Query.API, adapter: HomeBaked.MySQL.Adapter`, that’s why `Money.Ecto.Query.API` declares a behaviour in the first place.

But in general, we cannot implement it “for map,” because it heavily relies on how the map is mapped (pun unintended) into the database. For `mongo`, or `khepri`, we’ll need to [most likely] introduce another adapter.

I can add `MySQL` implementation into this PR, or we might split it into two. It would take a while because the last time I dealt with MySQL was in high school, but it should not take too long.

> And I think `money_is` would be more appropriately named `money_eq`

🤦🏼 of course, stay tuned.

**Comment by am-kantox** [comment]
That was fun! I never did as much SQL for decades.

Now we have three implementations of API,

- `use Money.Ecto.Query.API, adapter: Money.Ecto.Query.API.Composite`
- `use Money.Ecto.Query.API, adapter: Money.Ecto.Query.API.Map.Postgres`
- `use Money.Ecto.Query.API, adapter: Money.Ecto.Query.API.Map.MySQL`

The former is what I committed today morning. Two having `Map` in the name are for `MySQL` JSON and `Postgres` JSONB respectively. The interface stays the same.

---

I have tested `MySQL` implementation, and there is a test for it with `@moduletag skip: true` because proper testing against both databases would have increased the size of this PR twice.

Hope it works for you.

**Comment by kipcole9** [comment]
This is a amazing @am-kantox, really appreciated. I'll review after my morning coffee and aim to merge today and unless there are any unexpected issues we should be able to publish this weekend.

**Comment by am-kantox** [comment]
There is no rush, also, if you have a different vision of elegance in some approaches, please don’t hesitate to simply update the code; at the end of the day, it’s your library and it’s better to be consistent.

I care about the result and I absolutely do not care about my code being thrown out to the trash cam, I am not [unfortunately] in my twenties anymore :)

File: lib/money/ecto/money_ecto_query_api.ex [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+    Native implementation of how to `sum` several records having a field
+    of the type `Money` in the DB.
+
+    For `Postgres`, it delegates to the function on the composite type,
+      for other implementation it should return a `Ecto.Query.API.fragment/1`.
+    """
+    @macrocallback sum(Macro.t(), cast? :: boolean()) :: Macro.t()
+
+    @doc false
+    defmacro __using__([]),
```
Comment (am-kantox) [inline lib/money/ecto/money_ecto_query_api.ex:51]:
```suggestion
    @doc false
    defmacro __using__(opts \\ [])
    
    defmacro __using__([]),
```

**Comment by am-kantox** [comment]
Sounds great!

**Pull request merged**